File: Data Structures/doublyLinkedList.js
class Node {
	constructor(value){
		this.value = value
		this.next = null
		this.prev = null
	}
}

class DoublyLinkedList {
	constructor (){
		this.head = null
		this.tail = null
		this.len = 0
	}
	append(value){
		let newNode = new Node(value)
		if (this.len === 0) {
			this.head = newNode
			this.tail = newNode
		} else {
			newNode.prev = this.tail
			this.tail.next = newNode
			this.tail = newNode
		}
		this.len++
		return this
	}
	prepend(value){
		let newNode = new Node(value)
		if (this.len === 0) {
			this.head = newNode
			this.tail = newNode
		} else {
			newNode.next = this.head
			this.head.prev = newNode
			this.head = newNode
		}
		this.len++
		return this
	}
	insert(index,value){
		if (index < 0 || index > this.len) {
			return "Nothing to insert"
		}
		if (this.len === 0 || index === 0) {
			return this.prepend(value)
		}
		if (index === this.len) {
			return this.append(value)
		}
		let newNode = new Node(value)
		let leading = this.traverse(index -1)
		let following = leading.next
		newNode.next = following
		newNode.prev = leading
		leading.next = newNode
		following.prev = newNode
		this.len++
		return this
	}
	traverse(index){
		let curr = this.head
		let counter = 0
		while (counter < index) {
			curr = curr.next
			counter++
		}
		return curr
	}
	set(index,value){
		if (index < 0 || index >= this.len) {
			return "Nothing to set"
		}
		let curr = this.traverse(index)
		curr.value = value
		return this
	}
	remove(index){
		if (index < 0 || index >= this.len) {
			return "Nothing to remove"
		}
		let popped
		if (this.len === 1) {
			popped = this.head
			this.head = null
			this.tail = null
		} else if (index === 0) {
			popped = this.head
			this.head = this.head.next
			this.head.prev = null
		} else if (index === this.len-1) {
			popped = this.tail
			this.tail = this.tail.prev
			this.tail.next = null

		} else {
			let leading = this.traverse(index-1)
			popped = leading.next
			let following = popped.next
			leading.next = following
			following.prev = leading
		}

		this.len--
		return popped
	}
	isEmpty(){
		return Boolean(!this.len)
	}
	print(){
		let list = []
		let curr = this.head
		while(curr){
			list.push(curr.value)
			curr = curr.next
		}
		return console.log(list)
	}
}

let myDLL = new DoublyLinkedList()

File: Data Structures/doublyLinkedList.py
class Node:
	def __init__(self,value):
		self.value = value
		self.next = None
		self.prev = None

class DoublyLinkedList:
	def __init__(self):
		self.head = None
		self.tail = None
		self.length = 0

	def isEmpty(self):
		return bool(self.length is 0)

	def log(self):
		array = []
		curr = self.head
		while (curr):
			array.append(curr.value)
			curr = curr.next
		return print(array)

	def append(self, value):
		newNode = Node(value)
		if (self.length == 0):
			self.head = newNode
			self.tail = newNode
		else:
			newNode.prev = self.tail
			self.tail.next = newNode
			self.tail = newNode

		self.length += 1
		return self

	def prepend(self, value):
		newNode = Node(value)
		if (self.length == 0):
			self.head = newNode
			self.tail = newNode
		else:
			newNode.next = self.head
			self.head.prev = newNode
			self.head = newNode

		self.length += 1
		return self

	def insert(self, index, value):
		if (index < 0 or index > self.length):
			return "Nothing to insert"

		if (self.length == 0 or index == self.length):
			return self.append(value)

		if (index == 0):
			return self.prepend(value)

		newNode = Node(value)
		leading = self.traverse(index-1)
		following = leading.next

		newNode.prev = leading
		newNode.next = following
		leading.next = newNode
		following.prev = newNode
		self.length+=1
		return self

	def traverse(self, index):
		curr = self.head
		counter = 0
		while(counter < index):
			curr = curr.next
			counter+=1
		return curr

	def set(self, index, value):
		if (index < 0 or index >= self.length):
			return "Nothing to set"

		curr = self.traverse(index)
		curr.value = value
		return self

	def remove(self, index):
		if (index < 0 or index >= self.length):
			return "Nothing to remove"

		popped = ''
		if (self.length==1):
			popped = self.head
			self.head = None
			self.tail = None
		elif (index == 0):
			popped = self.head
			self.head = self.head.next
			self.head.prev = None
		elif (index == self.length-1):
			popped = self.tail
			self.tail = self.tail.prev
			self.tail.next = None
		else:
			leading = self.traverse(index-1)
			popped = leading.next
			following = popped.next

			leading.next = following
			following.prev = leading


		self.length-=1
		return popped

myNewDLL = DoublyLinkedList()